### PR TITLE
29 tcp chat

### PR DESCRIPTION
**Pull request by metaone**

### CODE REVIEW

File: 29_tcp_chat/main.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+func main() {
+	chat = &Chat{
+		join: make(chan net.Conn),
+		rooms: make(map[string]*ChatRoom),
+		clients: make(map[string]*Client),
+	}
+
+	chat.rooms[DEFAULT_ROOM] = CreateRoom()
+	chat.Start()
```
Comment (8tomat8) [inline 29_tcp_chat/main.go:158]: Call `go chat.Start()` will be more efficient than initialization of anon func in `chat.Start()` =)

File: 29_tcp_chat/main.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+	// start TCP
+	listener, _ := net.Listen("tcp", ":6666")
+	for {
+		conn, _ := listener.Accept()
```
Comment (8tomat8) [inline 29_tcp_chat/main.go:163]: If  `err != nil` you will get panic on hole application.

File: 29_tcp_chat/main.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		id: conn.RemoteAddr().String(),
+		username: conn.RemoteAddr().String(),
+		room: DEFAULT_ROOM,
+		input: make(chan string),
+		output: make(chan string),
+		reader: bufio.NewReader(conn),
+		writer: bufio.NewWriter(conn),
+	}
+
+	go client.Read()
```
Comment (arbrix) [inline 29_tcp_chat/main.go:93]: potential memory leaks, the control of ending those goroutines are absent

File: 29_tcp_chat/main.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+func (client *Client) Message(msg string) {
+	client.writer.WriteString(msg)
+	client.writer.Flush()
+}
+
+func CreateClient(conn net.Conn) *Client {
+	client := &Client{
+		id: conn.RemoteAddr().String(),
```
Comment (arbrix) [inline 29_tcp_chat/main.go]: here we see that only `RemoteAddr().String()` is used, no need to copy entire `conn` object

File: 29_tcp_chat/main.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+					client.Message(key + "\n")
+				}
+			case "/addroom":
+				if _, ok := chat.rooms[command[1]]; ok {
+					client.Message("Error: room already exists\n")
+				} else {
+					chat.rooms[command[1]] = CreateRoom()
+				}
+			case "/joinroom":
+				if _, ok := chat.rooms[command[1]]; ok {
```
Comment (arbrix) [inline 29_tcp_chat/main.go:54]: what if `command[1] == clinet.room`?

File: 29_tcp_chat/main.go (hunk 1 above)
Comment (arbrix) [reply 29_tcp_chat/main.go:158]: Disagree, in this case in `chat.Start()` launched the event loop and it runs at goroutine. The user of API is no need to know about that.

File: 29_tcp_chat/main.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	go client.Write()
+
+	return client
+}
+
+
+type ChatRoom struct {
+	clients map[string]*Client
+	input chan string
+	output chan string
```
Comment (arbrix) [inline 29_tcp_chat/main.go:103]: declared but not used

File: 29_tcp_chat/main.go (hunk 2 above)
Comment (arbrix) [reply 29_tcp_chat/main.go:163]: and where the logic for connection closing?

File: 29_tcp_chat/main.go (hunk 1 above)
Comment (8tomat8) [reply 29_tcp_chat/main.go:158]: Disagree with that it will be more efficient?)

**Pull request merged**